### PR TITLE
TRAVELI BACK END

### PR DESCRIPTION
**Pull request by carreraprogrammer**
The PR of the front-end part could be found [here](https://github.com/romans-adi/travel-booking-frontend/pull/15)

# Traveli Backend: Empowering Agencies to Manage Travel Experiences

Traveli's backend is the powerhouse that enables travel agencies to manage and offer their unique travel experiences to a global user base. This section delves into the essential features and technologies that power the backend of the Traveli application.

## Authentication and Authorization :bust_in_silhouette: :key:

- **Devise Gem**: The backend continues to rely on the Devise gem for user authentication, ensuring a secure and reliable registration and login process.
- **CanCanCan Gem**: Robust authorization is maintained with the CanCanCan gem, allowing fine-grained control over user roles and permissions.
- **User and Agency Roles**: Traveli maintains two distinct user roles through ability:
    I. **User**: Users can read, delete, and make reservations, but they cannot create travel listings. They also have access to travel and destination information.
    II. **Agency**: Agencies can read, delete, and create travel listings but are restricted from making reservations. They also have access to travel and destination information.

## Travel Management :airplane:

- **Add and Delete Travel Controller**: This exclusive functionality is available only for travel agencies.
    - **Add Travel**: A two-step process that empowers agencies to create travel proposals. First, agencies add detailed place information, including name, description, and image Next, they create compelling travel proposals, including pricing, trip duration,  group size, and travel type
    - **Delete Travel Page**: Agencies can efficiently manage their portfolio by removing outdated or unavailable travel listings.

## Places Controller

This controller handles the `API` endpoints related to places in the Traveli application. It provides `CRUD` operations for `places` and `includes authentication and authorization checks`.

Actions and Responsibilities:
  - `index`: Returns a list of all places.
  - `show`: Returns details of a specific place.
  - `create`: Creates a new place.
  - `place_params`: Private method to define strong parameters for creating a place.

## Reservations Controller
This controller is responsible for handling `API` endpoints related to `reservations`. It includes actions for `creating`, `listing`, `showing`, and `deleting` reservations.

Actions and Responsibilities:
  - `index`: Returns a list of reservations for the currently authenticated user.
  - `show`: Returns details of a specific reservation.
  - `create`: Creates a new reservation.
  - `destroy`: Deletes a reservation.
  - `reservation_params`: Private method to define strong parameters for creating a reservation.

## Travels Controller

The `Travels Controller` is responsible for managing `travel-related endpoints` in the Traveli application. It provides actions for `listing`, `showing`, `creating`, and `deleting` travel records.

Actions and Responsibilities:
  - `index`: Returns a list of all travel records.
  - `show`: Returns details of a specific travel record.
  - `create`: Creates a new travel record.
  - `destroy`: Deletes a travel record.
  - `travel_params`: Private method to define strong parameters for creating a travel record.

## Model Associations:

  - `Place`: The `Place` model has a `one-to-one association with the Travel model`, indicating that `each place can have one travel record associated with it`.

  - `Travel`: The `Travel` model belongs to a place and `has many users through reservations`. It also `has many reservations`. It includes `validations` for the `name` and `image fields`.

## CORS Configuration:

- The `CORS` configuration in `config/application.rb` ensures that the API `can accept cross-origin AJAX requests`. It allows `any origin` to access the API endpoints and specifies the allowed HTTP methods and headers.

### CODE REVIEW

File: app/controllers/api/v1/reservations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    else
+      render json: @reservation.errors, status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    @reservation = Reservation.find(params[:id])
+    @reservation.destroy
+    head :no_content
+  end
```
Comment (Meltrust) [inline app/controllers/api/v1/reservations_controller.rb:28]: - [ ] What happens if the reservation is not correctly destroyed? Kindly add some error handling. 👍

File: README.md [hunk 2]
```diff
@@ -1 +1,259 @@
-# travel-booking
\ No newline at end of file
+<!-- PROJECT DESCRIPTION -->
```
Comment (Meltrust) [inline README.md:1]: - [ ] OPTIONAL: Add some unit tests in order to preserve your app's integrity. 👍

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @carreraprogrammer ,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  Nice code organization ✔️ 
-  App is working well✔️ 
-  Good readme ✔️ 
-  Beautiful frontend ✔️ 



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

**Review by Meltrust — APPROVED** [review]
#### Hi @carreraprogrammer,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- Error handling has been added to controllers✔️ 
- Tests have been added✔️ 




Cheers and Happy coding!👏👏👏


------

**Comment by romans-adi** [comment]
Thank you for your review, @Meltrust! Wish you a jolly day! :wine_glass:

**Pull request merged**